src/rust/bitbox02/src/util.rs

```diff
@@ -78,6 +96,39 @@ macro_rules! str_to_cstr_force {
     };
 }
 
+/// Helper struct to convert from a list of Rust strings to a list of C strings of type `char**`.
+pub struct CStringList {
+    /// Buffers to hold the C strings, including null terminators.  This field
+    /// is never accessedread, but needs to be retained to keep the memory
+    /// pointed to by `words` valid.
+    _buffers: Vec<Vec<u8>>,
+    /// List of pointers to the C strings, pointing into `_buffers`.
+    strings: Vec<*const util::c_types::c_char>,
+}
+
+impl CStringList {
+    pub fn new(strings: Vec<String>) -> Result<CStringList, ()> {
+        let mut buffers: Vec<Vec<u8>> = Vec::with_capacity(strings.len());
```

Comment:
the `strings` list has 2048 bip39 words, all heap allocated. This then allocates another 2048*12 bytes = 24kb (12 bytes i the overhead of Vec<>). I sometimes run into a malloc error here. I guess it depends on the heap fragmentation and whether so many contiguous bytes can be found.


Turns out `Vec<u8>` for a word is 12 byte overhead with a maximum length of 8 chars (longest bip39 word), so less efficient than just doing a fixed `[u8; 9]`. Const generics would come in handy here...  

We have [96kb heap size](https://github.com/digitalbitbox/bitbox02-firmware/blob/29ae4d7a98622c5fcaeeb7f587ef45cbbb1e19e1/src/CMakeLists.txt#L551).

It was reduced from 128kb [here](https://github.com/digitalbitbox/bitbox02-firmware/pull/86) - @NickeZ do you remember why you reduced it?

Possible solutions I can think of:

1. Use `[u8; 9]` instead of `Vec<u8>` to save bytes. This sucks a bit as it special-purposes `CStringList`, but then again, it is only used for BIP39 words currently. Const generics would be nice now.
2. Shuffling 2048 words back and forth between C and Rust generates this overhead. I could keep the BIP39 words squarely in C-land (possibly only carrying a pointer through Rust).
3. Other ideas? 

@x1ddos

Reply:
I did number 2 in the commit `rust/trinary_input_string: use opaque type wrapping BIP39 wordlist`. It gets rid of all the CWords/CStringList business and has no needless memory overhead.

Reply:
I'm pretty sure I reduced it because we added more statically allocated buffers, like a separate 7kb one for u2f.

Reply:
> I did number 2 in the commit `rust/trinary_input_string: use opaque type wrapping BIP39 wordlist`. It gets rid of all the CWords/CStringList business and has no needless memory overhead.

Nice but I would now sqash it with some previous commits.

Reply:
done